internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,232 @@
+		params.EndTime = time.Now()
+	}
+
+	if params.StartTime.IsZero() {
+		params.StartTime = params.EndTime.Add(-time.Hour * 24 * 14) // default to 2 weeks ago
+	}
+}
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateInterval(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
+
+	msm := types.MinimumShouldMatch(1)
+	min := 1
+	filter := getHistoricalPrivilegeFilter(params.PrivilegeIDs)
+	return search.Request{
+		// get documents that match the device ID, are within the time range, and have a vehicle field
+		Query: &types.Query{
+			Bool: &types.BoolQuery{
+				Filter: []types.Query{
+					{Term: map[string]types.TermQuery{subjectField: {Value: params.DeviceID}}},
+					{Range: map[string]types.RangeQuery{timeField: types.DateRangeQuery{Gte: &startArg, Lte: &endArg}}},
+				},
+				Should: []types.Query{
+					{Exists: &types.ExistsQuery{Field: "data.vehicle"}},
+				},
+				MinimumShouldMatch: &msm,
+			},
+		},
+		Aggregations: map[string]types.Aggregations{
+			// docByInterval histogram aggregates the data into buckets based on the interval
+			docByInterval: {
+				DateHistogram: &types.DateHistogramAggregation{
+					Field:         ptr(timeField),
+					FixedInterval: &intervalArg,
+					MinDocCount:   &min,
+				},
+				Aggregations: map[string]types.Aggregations{
+					// singleDoc aggregation selects the first data point in each bucket
+					// that matches the desired filter
+					singleDoc: {
+						TopHits: &types.TopHitsAggregation{
+							Size:    ptr(1),
+							Source_: &filter,
+						},
+					},
+				},
+			},
+		},
+		Size: ptr(0), // only return the aggregations
+	}
+}
+
+// calculateInterval calculates the desired interval in milliseconds. This interval will provide n buckets between the start and end times.
+// if the interval is 0 or negative it will be clamped to 1ms.
+func calculateInterval(start, end time.Time, buckets int) int {
+	duration := int(end.Sub(start).Milliseconds())
+	interval := duration / buckets
+	if interval < 1 {
+		return 1 // 0 or negative interval is not supported, clamp to 1
+	}
+	return interval
+}
+
+// getHitsFromHistoryResponse extracts the hits from the response body.
+// The maxValues parameter is used to limit the number of values returned.
+func getHitsFromHistoryResponse(body []byte, maxValues int) []any {
+	retData := []any{}
+	bucketKey := fmt.Sprintf("aggregations.%s.buckets", docByInterval)
+	hitsKey := fmt.Sprintf("%s.hits.hits", singleDoc)
+	gjson.GetBytes(body, bucketKey).ForEach(func(key, v gjson.Result) bool {
+		gjson.Get(v.Raw, hitsKey).ForEach(func(key, value gjson.Result) bool {
```

Comment:
@Allyson-English I assume you wrote this logic—what about this?

https://go.dev/play/p/jlOEauUPOfn

Reply:
But do I want to do this at all? I feel like we're parsing this thing when we do `.Value()`. Can we `json.RawMessage` this somehow?

Reply:
We should be able to use RAW since this output is eventually converted back to JSON when it's sent in the response.

Reply:
I think gjson is a relic of wanted to maniuplate the data instead of just passing it through... are you thinking something like this? https://go.dev/play/p/zOle8BP1YNy

Reply:
Here are two versions of this 
1. Uses the typed Response object from elastic 
2. Uses gjson without allocations similar to what @Allyson-English provided

The gjson should be a bit performant since we don't have to unmarshal from elastic then remarshal for sending the response.
``` golang
// getHitsFromHistoryResponse extracts the hits from the response body.
// The maxValues parameter is used to limit the number of values returned.
func getHitsFromHistoryResponse(body []byte, maxValues int) []json.RawMessage {
	retData := []json.RawMessage{}
	// This key gets the first data point in each bucket which should only contain 1 value
	fullKey := fmt.Sprintf("aggregations.%s.buckets.#.%s.hits.hits.0._source", docByInterval, singleDoc)
	gjson.GetBytes(body, fullKey).ForEach(func(key, value gjson.Result) bool {
		// if we have reached the max number of values, stop
		if len(retData) == maxValues {
			return false
		}
		retData = append(retData, resultToBytes(body, value))
		return true
	})
	return retData
}

// resultToBytes converts the result to a byte slice.
// this logic is the recommended way to extract the raw bytes from a gjson result.
// more info: https://github.com/tidwall/gjson/blob/6ee9f877d683381343bc998c137339c7ae908b86/README.md#working-with-bytes
func resultToBytes(body []byte, result gjson.Result) []byte {
	var raw []byte
	if result.Index > 0 {
		raw = body[result.Index : result.Index+len(result.Raw)]
	} else {
		raw = []byte(result.Raw)
	}
	return raw
}

// getHitsFromHistoryResponseTyped extracts the hits from the response body.
func getHitsFromHistoryResponseTyped(resp *search.Response, maxValues int) ([]json.RawMessage, error) {
	retData := []json.RawMessage{}
	datAgg, ok := resp.Aggregations[docByInterval].(*types.DateHistogramAggregate)
	if !ok {
		return retData, fmt.Errorf("failed to get data from response")
	}
	buckets, ok := datAgg.Buckets.([]types.DateHistogramBucket)
	if !ok {
		return retData, fmt.Errorf("failed to get data from response")
	}
	for _, bucket := range buckets {
		singAgg, ok := bucket.Aggregations[singleDoc].(*types.TopHitsAggregate)
		if !ok {
			return retData, fmt.Errorf("failed to get data from response")
		}
		for _, hit := range singAgg.Hits.Hits {
			if len(retData) >= maxValues {
				return retData, nil
			}
			retData = append(retData, hit.Source_)
		}
	}

	return retData, nil
}

```

Reply:
`❤️ == gjson` All the casting when using the typed response hurts.